Fix Internal Server Error on Wrong Password

andy-dendy-repo opened the pull request:
Auth with wrong password gives an Internal Error. If accountCredentials is null, return Unauthorized error.

andy-dendy-repo committed:
Fix Internal Server Error on Wrong Password

Hlib-mats commented on CharlieBackend.Business/Services/AccountService.cs:
```diff
@@ -72,7 +72,14 @@ public async Task<Result<AccountDto>> GetAccountCredentialsAsync(AuthenticationD
                 authenticationModel.Password = PasswordHelper.HashPassword(authenticationModel.Password, salt);
 
-                var foundAccount = _mapper.Map<AccountDto>(await _unitOfWork.AccountRepository.GetAccountCredentials(authenticationModel));
+                var cred = await _unitOfWork.AccountRepository.GetAccountCredentials(authenticationModel);
```
please rename cred to accountCredentials for better understanding(code conventions)

Hlib-mats commented on CharlieBackend.Business/Services/AccountService.cs:
```diff
@@ -72,7 +72,14 @@ public async Task<Result<AccountDto>> GetAccountCredentialsAsync(AuthenticationD
+
+                if(cred == null)
+                {
+                    return Result<AccountDto>.GetError(ErrorCode.Unauthorized, "Email or password is incorrect.");
```
Error should be NotFound I think, because if credentials is null, there is no accounts that matches

andy-dendy-repo replied:
credentials can be null if password is wrong and email exists

andy-dendy-repo committed:
Update AccountService.cs

andy-dendy-repo committed:
Update AccountService.cs

QuestroL commented on CharlieBackend.Business/Services/AccountService.cs:
```diff
@@ -72,7 +72,14 @@ public async Task<Result<AccountDto>> GetAccountCredentialsAsync(AuthenticationD
+
+                if(accountCredentials == null)
+                {
+                    return Result<AccountDto>.GetError(ErrorCode.Unauthorized, "Email or password is incorrect.");
```
Have you deleted the same validation in Account Controller?

andy-dendy-repo replied:
Validation layer is implemented by Business Services. I don't see the same validation in AccontController.

QuestroL replied:
This code you wiil see in Account Controller  in SingIn method .I think it will be better if you move Validation  for IsActive and delete (foundAccount == null).       
 if (foundAccount == null)
            {
                return Unauthorized("Incorrect credentials, please try again.");
            }

            if (!foundAccount.IsActive)
            {
                return StatusCode(401, "Account is not active!");
            }

andy-dendy-repo replied:
We've decided to leave this for further development, when all validation will be placed in extra layer.

gingerbread-monster commented on CharlieBackend.Business/Services/AccountService.cs:
```diff
@@ -72,7 +72,14 @@ public async Task<Result<AccountDto>> GetAccountCredentialsAsync(AuthenticationD
+
+                var foundAccount = _mapper.Map<AccountDto>(accountCredentials);
 
                 return Result<AccountDto>.GetSuccess(foundAccount);
```
Currently, in **3b2e3b6b3d882894f5a6d95eca7268664518197e** (your latest commit by now), this line here is unreachable

gingerbread-monster commented:
`AccountService.GetAccountCredentialsAsync()` requires some logic refactoring.

We make 2 database calls instead of one.

First is inside `AccountRepository.GetAccountSaltByEmail()`.
It makes a call to database, and after **ALREADY** finding user account by email it just throws it away and returns only salt.

Second is inside `AccountRepository.GetAccountCredentials()`.
It makes a database call again using both email and password.

Please update method logic.

andy-dendy-repo committed:
Update AccountService.cs

Code refactoring with respect to https://github.com/ita-social-projects/WhatBackend/pull/331#issuecomment-768534981

gingerbread-monster commented:
Also, I think [this comment](https://github.com/ita-social-projects/WhatBackend/pull/331#discussion_r562617952) here suggests removing this check
https://github.com/ita-social-projects/WhatBackend/blob/0f78082fd6c8f03ede68ab971f5a5d0371028fc6/CharlieBackend.Api/Controllers/AccountsController.cs#L73

which is useless now since you already added it to `AccountService.GetAccountCredentialsAsync()` with your changes.

andy-dendy-repo commented:
> Also, I think [this comment](https://github.com/ita-social-projects/WhatBackend/pull/331#discussion_r562617952) here suggests removing this check
> https://github.com/ita-social-projects/WhatBackend/blob/0f78082fd6c8f03ede68ab971f5a5d0371028fc6/CharlieBackend.Api/Controllers/AccountsController.cs#L73
> 
> which is useless now since you already added it to `AccountService.GetAccountCredentialsAsync()` with your changes.

This check is necessary to return described Error. Without it server returns Internal Error on wrong credentials.
